controllers/provider-gcp/pkg/validator/shoot_validator.go

```diff
@@ -0,0 +1,157 @@
+
+	if errList := gcpvalidation.ValidateNetworking(shoot.Spec.Networking, networkPath); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	// Provider validation
+	fldPath := field.NewPath("spec", "provider")
+
+	// InfrastructureConfig
+	infraConfigFldPath := fldPath.Child("infrastructureConfig")
+
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.Required(infraConfigFldPath, "InfrastructureConfig must be set for GCP shoots")
+	}
+
+	infraConfig, err := decodeInfrastructureConfig(v.decoder, shoot.Spec.Provider.InfrastructureConfig, infraConfigFldPath)
+	if err != nil {
+		return err
+	}
+
+	if errList := gcpvalidation.ValidateInfrastructureConfig(infraConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	cloudProfile := &gardencorev1beta1.CloudProfile{}
+	if err := v.client.Get(ctx, kutil.Key(shoot.Spec.CloudProfileName), cloudProfile); err != nil {
+		return err
+	}
+
+	controlPlaneConfigPath := fldPath.Child("controlPlaneConfig")
+	if shoot.Spec.Provider.ControlPlaneConfig != nil {
+		controlPlaneConfig, err := decodeControlPlaneConfig(v.decoder, shoot.Spec.Provider.ControlPlaneConfig, controlPlaneConfigPath)
+		if err != nil {
+			return err
+		}
+
+		if errList := gcpvalidation.ValidateControlPlaneConfig(controlPlaneConfig, shoot.Spec.Region, cloudProfile.Spec.Regions); len(errList) != 0 {
+			return errList.ToAggregate()
+		}
+	}
+
+	var (
+		workersPath  = fldPath.Child("workers")
+		allowedZones = getAllowedRegionZonesFromCloudProfile(shoot, cloudProfile)
+	)
+
+	if errList := gcpvalidation.ValidateWorkers(shoot.Spec.Provider.Workers, allowedZones, workersPath); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	return nil
+}
+
+func (v *Shoot) validateShootUpdate(ctx context.Context, oldShoot, shoot *garden.Shoot) error {
+	fldPath := field.NewPath("spec", "provider")
+
+	// InfrastructureConfig
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.Required(fldPath.Child("infrastructureConfig"), "InfrastructureConfig must be set for GCP shoots")
+	}
+
+	infraConfig, err := decodeInfrastructureConfig(v.decoder, shoot.Spec.Provider.InfrastructureConfig, fldPath)
+	if err != nil {
+		return err
+	}
+
+	if oldShoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.InternalError(fldPath.Child("infrastructureConfig"), errors.New("InfrastructureConfig is not available on old shoot"))
+	}
+
+	oldInfraConfig, err := decodeInfrastructureConfig(v.decoder, oldShoot.Spec.Provider.InfrastructureConfig, fldPath)
+	if err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(oldInfraConfig, infraConfig) {
+		if errList := gcpvalidation.ValidateInfrastructureConfigUpdate(oldInfraConfig, infraConfig); len(errList) != 0 {
+			return errList.ToAggregate()
+		}
+	}
+
+	// ControlPlaneConfig
+	if shoot.Spec.Provider.ControlPlaneConfig == nil {
```

Comment:
For an update `shoot.Spec.Provider.ControlPlaneConfig` must be set whereas it's allowed to leave it out for creation events?